### ISSUE TITLE
Consuming can skip messages

### ISSUE THREAD

**Issue by tmasternak** [NServiceBusExtensions/NServiceBus.Native#2]
## Overview
The current approach to looping over messages can result in messages being skipped and never processed.

## Details
The looping logic is based on the `RowVersion` column value. On each loop step we read number of messages and record the highest seen value in that column. On next spin we select only those messages which have higher `RowVersion` value.

The statement used for consumption uses `select .. with (readpast)` which skips any rows for which there are locks held by some running transaction. In a situation when we have a single transaction that takes longer to commit  followed by a transaction that gets committed the row from slow transaction will be skipped and never consumed again.

**Comment by SimonCropp** [comment]
any suggestions on how to fix it?

**Comment by tmasternak** [comment]
TL;DR Using `RowVersion` won't work we need a single, async sequencer that will put valid row numbers.

### Longer answer
I think that relying on `RowVersion` set by each write transaction will be hard to get to work. The possibilities there boil down to isolation level of the reader. So we can do:
 * `select ... with (readpast)` - the problems were already mentioned,
 * `select ... with (nolock)` - we can read uncommited changes and process ghost messages - bad
 * `select ... ` - the reader has to wait for all locks on the table which will probably make performance of the reader unacceptable especially under high write load.

What could work is a separate sequencer + addintional nullable column. Say we had `MessageNo INT NULL` . All writes are unchanged and the reader behaves just like now (read by messageNo and keep the highest value seen so far). 

This leaves the question of the sequencer operations. First of all it probably has to be a singleton to make sure we can have reasonable assumptions on values of MessageNo (otherwise we are back in the place of multiple writers). Secondly it probably should be part of each receiver to make the hosting easier. Considering both it means that we need some coordination between sequencers for competing consumers at the receiver to make sure only one is putting `MessageNo`. With sql this can probably work with shared lock on configuration table.

**Comment by SzymonPobiega** [comment]
@tmasternak you mean a sequencer would assign a sequence to a bunch of messages and in the same time touch the lock raw to ensure it is still the master of the sequencing?

**Comment by tmasternak** [comment]
@SzymonPobiega yes.  The pseudo code for each sequencer could be something like:

```
  while(true) 
  {
      begin transaction;
      update config_row set round = round +1 with where id = const_id (readpast);
      if(affected_row == 0) {
        end transaction;
        back_off;
      }
      sequence_batch_of_rows;
      commit transaction;
  }
```

**Comment by SzymonPobiega** [comment]
Tomek, another approach might be to query with readpast and if there are
any holes in the retrieved batch, try inserting dummy messages there with
no wait in the same transaction as updating the index

**Comment by SimonCropp** [comment]
@SzymonPobiega interesting approach.

**Comment by SzymonPobiega** [comment]
@SimonCropp I've been bitten by this problem in my previous life when implementing event sourcing on top of an Oracle DB. Once every couple millions of records my denormalizers would skip ove an initial event in the payment process which made all the subsequent event fail in processing.

We were lucky that in our case the hole in the sequence was caused by a transaction that took 30 seconds (instead of an average < 1s) to execute (GC pause) so we were able to pinpoint these cases by reading the logs. It the cause was less obvious (i.e. 3 s instead of 1 s) it would be much harder.

**Comment by SimonCropp** [comment]
@tmasternak @SzymonPobiega do one of you want to have a go at a PR?
